fuse/src/main/java/tachyon/fuse/TachyonFuseFs.java

```diff
@@ -0,0 +1,577 @@
+      return -ErrorCodes.EBADFD();
+    }
+    if (oe.mOut == null) {
+      LOG.error("{} was not open for writing", path);
+      return -ErrorCodes.EBADFD();
+    }
+    try {
+      oe.mOut.flush();
+    } catch (IOException e) {
+      return -ErrorCodes.EIO();
+    }
+
+    return 0;
+  }
+
+  @Override
+  public int getattr(String path, FileStat stat) {
+    int ret = 0;
+    final TachyonURI turi = mPathResolverCache.getUnchecked(path);
+    LOG.trace("getattr({}) [Tachyon: {}]", path, turi);
+    try {
+      final TachyonFile tf = mTFS.openIfExists(turi);
+      if (tf == null) {
+        ret = ErrorCodes.ENOENT();
+      } else {
+        final FileInfo fi = mTFS.getInfo(tf);
+        stat.st_size.set(fi.getLength());
+
+        final long ctime = fi.getLastModificationTimeMs();
+        final long ctime_sec = fi.getLastModificationTimeMs() / 1000;
+        final long ctime_nsec = (fi.getLastModificationTimeMs() % 1000) * 1000;
+        stat.st_ctim.tv_sec.set(ctime_sec);
+        stat.st_ctim.tv_nsec.set(ctime_nsec);
+        stat.st_mtim.tv_sec.set(ctime_sec);
+        stat.st_mtim.tv_nsec.set(ctime_nsec);
+
+        System.getProperty("user.name");
+
+        stat.st_uid.set(UID_AND_GID[0]);
+        stat.st_gid.set(UID_AND_GID[1]);
+
+        final int mode;
+        if (fi.isFolder) {
+          mode = FileStat.S_IFDIR;
+        } else {
+          mode = FileStat.S_IFREG;
+        }
+        stat.st_mode.set(mode);
+      }
+
+    } catch (InvalidPathException e) {
+      LOG.debug("Invalid path {}", path, e);
+      ret = ErrorCodes.ENOENT();
+    } catch (FileDoesNotExistException e) {
+      LOG.debug("File does not exist {}", path, e);
+      ret = ErrorCodes.ENOENT();
+    } catch (IOException e) {
+      LOG.error("IOException on {}", path, e);
+      ret = ErrorCodes.EFAULT();
+    } catch (TachyonException e) {
+      LOG.error("TachyonException on {}", path, e);
+      ret = ErrorCodes.EFAULT();
+    } catch (Throwable e) {
+      LOG.error("Unexpected exception on {}", path, e);
+      ret = ErrorCodes.EFAULT();
+    }
+
+    return -ret;
+  }
+
+  @Override
+  public String getFSName() {
+    return "tachyon-fuse";
+  }
+
+  @Override
+  public int mkdir(String path, @mode_t long mode) {
+    int ret = 0;
+    final TachyonURI turi = mPathResolverCache.getUnchecked(path);
+    LOG.trace("mkdir({}) [Tachyon: {}]", path, turi);
+    LOG.warn("{}: mode is ignored in tachyon-fuse", path);
+    try {
+      mTFS.mkdir(turi);
+    } catch (FileAlreadyExistsException e) {
+      LOG.debug("Cannot make dir. {} already exists", path, e);
+      ret = ErrorCodes.EEXIST();
+    } catch (InvalidPathException e) {
+      LOG.debug("Cannot make dir. Invalid path: {}", path, e);
+      ret = ErrorCodes.ENOENT();
+    } catch (IOException e) {
+      LOG.error("Cannot make dir. IOException: {}", path, e);
+      ret = ErrorCodes.EIO();
+    } catch (TachyonException e) {
+      LOG.error("Cannot make dir. {}", path, e);
```

Comment:
Perhaps we can give different error codes based on the specific type of TachyonException.